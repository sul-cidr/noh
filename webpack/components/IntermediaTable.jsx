import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import IntermediaTitle from "./IntermediaTitle";
import IntermediaElement from "./IntermediaElement";

class IntermediaTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentSectionIndex: 0
    };
  }

  componentWillReceiveProps() {
    // Whenever props change, e.g. time, check to see if the current section index has changed. If so, update state.
    const currentSectionIndex = this.findCurrentSectionIndex();
    if (currentSectionIndex !== this.state.currentSectionIndex) {
      this.updateIndex(currentSectionIndex);
    }
  }

  findCurrentSectionIndex() {
    // Determine current sectionIndex based on time
    const currentSectionIndex = this.props.sections.findIndex(
      section =>
        this.props.currentTime >= section.startTime.value &&
        this.props.currentTime < section.endTime.value
    );
    // since findIndex returns -1 if it finds nothing, just return the current section index from state
    if (currentSectionIndex === -1) {
      return this.state.currentSectionIndex;
    }
    return currentSectionIndex;
  }

  updateIndex(currentSectionIndex) {
    this.setState({ currentSectionIndex });
  }

  render() {
    // due to lack of data for start and end time, not always finding current section
    const section = this.props.sections[this.state.currentSectionIndex];
    return (
      <div className="intermedia-table">
        <IntermediaTitle
          section={section.sectionName.value}
          play={this.props.play}
        />
        <IntermediaElement
          fieldName="Voices"
          fieldValue={section.numberVoices.value}
        />
        <IntermediaElement
          fieldName="Type of voice"
          fieldValue={section.voice.value}
        />
        <IntermediaElement fieldName="Text" fieldValue={section.text.value} />
        <IntermediaElement
          fieldName="Percussion"
          fieldValue={section.percussion.value}
        />
        <IntermediaElement
          fieldName="Type of percussion"
          fieldValue={section.numberOfPercussion.value}
        />
        <IntermediaElement
          fieldName="Nohkan"
          fieldValue={section.nokhanPresent.value}
        />
        <IntermediaElement
          fieldName="Dance"
          fieldValue={section.dancePresent.value}
        />
      </div>
    );
  }
}

IntermediaTable.propTypes = {
  play: PropTypes.string.isRequired,
  currentTime: PropTypes.number.isRequired,
  sections: PropTypes.arrayOf(
    PropTypes.shape({
      play: PropTypes.shape({ value: PropTypes.string }),
      sectionName: PropTypes.shape({ value: PropTypes.string }),
      intensity: PropTypes.shape({ number: PropTypes.string }),
      startTime: PropTypes.shape({ value: PropTypes.number }),
      endTime: PropTypes.shape({ value: PropTypes.number }),
      numberVoices: PropTypes.shape({
        number: PropTypes.string,
        value: PropTypes.string
      }),
      voice: PropTypes.shape({ value: PropTypes.string }),
      text: PropTypes.shape({ value: PropTypes.string }),
      numberOfPercussion: PropTypes.shape({ number: PropTypes.string }),
      percussion: PropTypes.shape({ value: PropTypes.string }),
      nokhanPresent: PropTypes.shape({
        present: PropTypes.string,
        value: PropTypes.string
      }),
      dancePresent: PropTypes.shape({
        present: PropTypes.string,
        value: PropTypes.string
      }),
      captions: PropTypes.Array,
      narrative: PropTypes.shape({ value: PropTypes.string })
    })
  ).isRequired
};

const mapStateToProps = state => ({ currentTime: state.currentTime });

export const Unwrapped = IntermediaTable;
export default connect(mapStateToProps)(IntermediaTable);
