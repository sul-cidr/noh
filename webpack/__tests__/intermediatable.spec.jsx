import React from "react";
import { shallow, mount } from "enzyme";
import configureMockStore from "redux-mock-store";
import IntermediaTable, {
  Unwrapped as UnwrappedIntermediaTable
} from "../components/IntermediaTable";

const sections = [
  {
    sectionName: {
      value: "Nanori"
    },
    intensity: { number: "5" },
    startTime: { value: 0 },
    endTime: { value: 180 },
    numberVoices: { value: "Waki/Waki Tsure" },
    voice: { value: "Sung - Non Congruent" },
    text: { value: "Sashinori - Non Congruent" },
    numberOfPercussion: { value: "Otsuzumi + Kotsuzumi + Taiko" },
    percussion: { value: "Non Congruent" },
    nokhanPresent: { value: "Non Congruent" },
    dancePresent: { value: "Dance to text" }
  },
  {
    sectionName: {
      value: "Mondo"
    },
    intensity: { number: "5" },
    startTime: { value: 181 },
    endTime: { value: 300 },
    numberVoices: { value: "Waki/Waki Tsure" },
    voice: { value: "Sung - Non Congruent" },
    text: { value: "Sashinori - Non Congruent" },
    numberOfPercussion: { value: "Otsuzumi + Kotsuzumi + Taiko" },
    percussion: { value: "Non Congruent" },
    nokhanPresent: { value: "Non Congruent" },
    dancePresent: { value: "Dance to text" }
  }
];

describe("<UnwrappedIntermediaTable>", () => {
  it("renders as expected", () => {
    const component = shallow(
      <UnwrappedIntermediaTable
        play="hashitomi"
        sections={sections}
        currentTime={10}
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("correctly determines current section", () => {
    const component = shallow(
      <UnwrappedIntermediaTable
        play="hashitomi"
        sections={sections}
        currentTime={200}
      />
    );
    const currentSectionIndex = component.instance().findCurrentSectionIndex();
    expect(currentSectionIndex).toBe(1);
  });
  it("handles failing to find a section correctly", () => {
    const component = shallow(
      <UnwrappedIntermediaTable
        play="hashitomi"
        sections={sections}
        currentTime={5000}
      />
    );
    const currentSectionIndex = component.instance().findCurrentSectionIndex();
    expect(currentSectionIndex).toBe(0);
  });
  it("updates state correctly on componentWillReceiveProps", () => {
    const component = mount(
      <UnwrappedIntermediaTable
        play="hashitomi"
        sections={sections}
        currentTime={0}
      />
    );
    component.setProps({ currentTime: 200 });
    component.instance().componentWillReceiveProps();
    expect(component.state().currentSectionIndex).toBe(1);
  });
});

describe("<InterMediaTable>", () => {
  it("renders as expected while connected", () => {
    const initialState = { currentTime: 10 };
    const mockStore = configureMockStore();
    const store = mockStore(initialState);
    const wrapper = shallow(
      <IntermediaTable play="hashitomi" sections={sections} />,
      { context: { store } }
    );
    expect(wrapper).toMatchSnapshot();
  });
});
