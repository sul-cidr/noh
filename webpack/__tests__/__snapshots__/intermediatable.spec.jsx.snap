// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<InterMediaTable> renders as expected while connected 1`] = `
<IntermediaTable
  currentTime={10}
  dispatch={[Function]}
  play="hashitomi"
  sections={
    [
      {
        "dancePresent": {
          "value": "Dance to text",
        },
        "endTime": {
          "value": 180,
        },
        "intensity": {
          "number": "5",
        },
        "nokhanPresent": {
          "value": "Non Congruent",
        },
        "numberOfPercussion": {
          "value": "Otsuzumi + Kotsuzumi + Taiko",
        },
        "numberVoices": {
          "value": "Waki/Waki Tsure",
        },
        "percussion": {
          "value": "Non Congruent",
        },
        "sectionName": {
          "value": "Nanori",
        },
        "sectionUrl": "kokaji/nanori",
        "startTime": {
          "value": 0,
        },
        "text": {
          "value": "Sashinori - Non Congruent",
        },
        "voice": {
          "value": "Sung - Non Congruent",
        },
      },
      {
        "dancePresent": {
          "value": "Dance to text",
        },
        "endTime": {
          "value": 300,
        },
        "intensity": {
          "number": "5",
        },
        "nokhanPresent": {
          "value": "Non Congruent",
        },
        "numberOfPercussion": {
          "value": "Otsuzumi + Kotsuzumi + Taiko",
        },
        "numberVoices": {
          "value": "Waki/Waki Tsure",
        },
        "percussion": {
          "value": "Non Congruent",
        },
        "sectionName": {
          "value": "Mondo",
        },
        "sectionUrl": "kokaji/mondo",
        "startTime": {
          "value": 181,
        },
        "text": {
          "value": "Sashinori - Non Congruent",
        },
        "voice": {
          "value": "Sung - Non Congruent",
        },
      },
    ]
  }
/>
`;

exports[`<UnwrappedIntermediaTable> renders as expected 1`] = `
<div
  className="intermedia-table"
>
  <IntermediaTitle
    section="Nanori"
    sectionUrl="kokaji/nanori"
  />
  <IntermediaElement
    fieldName="Voices"
    fieldValue="Waki/Waki Tsure"
  />
  <IntermediaElement
    fieldName="Type of voice"
    fieldValue="Sung - Non Congruent"
  />
  <IntermediaElement
    fieldName="Text"
    fieldValue="Sashinori - Non Congruent"
  />
  <IntermediaElement
    fieldName="Percussion"
    fieldValue="Non Congruent"
  />
  <IntermediaElement
    fieldName="Type of percussion"
    fieldValue="Otsuzumi + Kotsuzumi + Taiko"
  />
  <IntermediaElement
    fieldName="Nohkan"
    fieldValue="Non Congruent"
  />
  <IntermediaElement
    fieldName="Dance"
    fieldValue="Dance to text"
  />
</div>
`;
